File: backend/Bingol-University-Social-Media-main/src/main/java/bingol/campus/story/business/abstracts/StoryService.java
package bingol.campus.story.business.abstracts;

import bingol.campus.followRelation.core.exceptions.BlockingBetweenStudent;
import bingol.campus.post.core.response.CommentDetailsDTO;
import bingol.campus.post.core.response.LikeDetailsDTO;
import bingol.campus.response.DataResponseMessage;
import bingol.campus.response.ResponseMessage;
import bingol.campus.story.core.exceptions.StoryNotActiveException;
import bingol.campus.story.core.exceptions.*;
import bingol.campus.story.core.response.FeatureStoryDTO;
import bingol.campus.story.core.response.StoryDTO;
import bingol.campus.story.core.response.StoryDetails;
import bingol.campus.student.core.response.SearchAccountDTO;
import bingol.campus.student.exceptions.*;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import javax.xml.crypto.Data;
import java.io.IOException;
import java.util.List;
import java.util.UUID;

public interface StoryService {
    ResponseMessage add(String username, MultipartFile file) throws StudentNotFoundException, IOException, OnlyPhotosAndVideosException, PhotoSizeLargerException, VideoSizeLargerException, FileFormatCouldNotException;

    ResponseMessage delete(String username, UUID storyId) throws StoryNotFoundException, StudentNotFoundException, OwnerStoryException;


    ResponseMessage featureStory(String username, UUID storyId,UUID featuredStoryId) throws StudentNotFoundException, StoryNotFoundException, OwnerStoryException, AlreadyFeaturedStoriesException, FeaturedStoryGroupNotFoundException;


    ResponseMessage extendStoryDuration(String username, UUID storyId, int hours) throws StudentNotFoundException, OwnerStoryException, StoryNotActiveException, InvalidHourRangeException, FeaturedStoryModificationException;

    List<SearchAccountDTO> getStoryViewers(String username, UUID storyId) throws StudentNotFoundException, OwnerStoryException, StoryNotActiveException;

    int getStoryViewCount(String username, UUID storyId) throws StudentNotFoundException, OwnerStoryException, StoryNotActiveException;

    DataResponseMessage<List<StoryDTO>> getPopularStories(String username) throws StudentNotFoundException;

    DataResponseMessage<List<StoryDTO>> getUserActiveStories(String username, Long userId) throws StudentNotFoundException, BlockingBetweenStudent, NotFollowingException;

    DataResponseMessage<List<CommentDetailsDTO>> getStoryComments(String username, UUID storyId) throws StudentNotFoundException, OwnerStoryException, StoryNotActiveException;

    DataResponseMessage<StoryDTO> viewStory(String username, UUID storyId) throws StoryNotFoundException, StoryNotActiveException, StudentNotFoundException, NotFollowingException, BlockingBetweenStudent;



    DataResponseMessage<List<LikeDetailsDTO>> getLike(String username, UUID storyId) throws StudentNotFoundException, OwnerStoryException, StoryNotActiveException;


    DataResponseMessage<List<StoryDetails>> getStories(String username, Pageable pageable) throws StudentNotFoundException;


    DataResponseMessage<StoryDetails> getStoryDetails(String username, UUID storyId, Pageable pageable) throws StudentNotFoundException, StoryNotFoundException, OwnerStoryException;

    ResponseMessage featureUpdate(String username, UUID featureId, String title, MultipartFile file) throws StudentNotFoundException, FeaturedStoryGroupNotFoundException, FeaturedStoryGroupNotAccess, IOException, OnlyPhotosAndVideosException, PhotoSizeLargerException, VideoSizeLargerException, FileFormatCouldNotException;

    DataResponseMessage<FeatureStoryDTO> getFeatureId(String username, UUID featureId) throws StudentNotFoundException, FeaturedStoryGroupNotFoundException, BlockingBetweenStudent, StudentProfilePrivateException;

    DataResponseMessage<List<FeatureStoryDTO>> getFeaturedStoriesByStudent(String username, Long studentId) throws StudentNotFoundException, BlockingBetweenStudent, StudentProfilePrivateException;

    DataResponseMessage<List<FeatureStoryDTO>> getMyFeaturedStories(String username) throws StudentNotFoundException;

    DataResponseMessage<List<StoryDTO>> archivedStories(String username) throws StudentNotFoundException;

    ResponseMessage deleteArchived(String username, UUID storyId);
}

File: frontend/campus_app/lib/services/storyService.dart
import 'package:dio/dio.dart';
import 'package:social_media/models/data_response_message.dart';
import 'package:social_media/models/response_message.dart';
import 'package:social_media/models/story_dto.dart';
import 'package:social_media/models/feature_story_dto.dart';
import 'package:social_media/models/story_details.dart';
import 'package:social_media/models/comment_details_dto.dart';
import 'package:social_media/models/search_account_dto.dart';

class StoryService {
  final Dio _dio;

  StoryService(this._dio) {
    _dio.options.baseUrl = 'http://localhost:8080/v1/api'; // Base URL ayarlandı
  }

  // Yeni hikaye ekleme
  Future<ResponseMessage> add(String token, MultipartFile file) async {
    final response = await _dio.post(
      '/story/add',
      data: FormData.fromMap({
        'file': file,
      }),
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Mevcut bir hikayeyi silme
  Future<ResponseMessage> delete(String token, String storyId) async {
    final response = await _dio.delete(
      '/story/$storyId',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Hikaye öne çıkarma
  Future<ResponseMessage> featureStory(String token, String storyId, String? featuredStoryId) async {
    final response = await _dio.put(
      '/story/$storyId/feature',
      queryParameters: {'featuredStoryId': featuredStoryId},
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Hikaye güncelleme
  Future<ResponseMessage> featureUpdate(String token, String featureId, String? title, MultipartFile? file) async {
    final response = await _dio.put(
      '/story/$featureId/update',
      data: FormData.fromMap({
        'title': title,
        'file': file,
      }),
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Belirli bir hikaye detaylarını alma
  Future<DataResponseMessage<StoryDetails>> getStoryDetails(String token, String storyId) async {
    final response = await _dio.get(
      '/story/$storyId',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<StoryDetails>.fromJson(
      response.data,
      (data) => StoryDetails.fromJson(data as Map<String, dynamic>),
    );
  }

  // Belirli bir öğrencinin öne çıkarılan hikayelerini getir
  Future<DataResponseMessage<List<FeatureStoryDTO>>> getFeaturedStoriesByStudent(String token, int studentId) async {
    final response = await _dio.get(
      '/story/student/$studentId/featured-stories',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<FeatureStoryDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => FeatureStoryDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Kullanıcının kendi öne çıkarılan hikayelerini getir
  Future<DataResponseMessage<List<FeatureStoryDTO>>> getMyFeaturedStories(String token) async {
    final response = await _dio.get(
      '/story/me/featured-stories',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<FeatureStoryDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => FeatureStoryDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Tüm hikayeleri sayfalı olarak listeleme
  Future<DataResponseMessage<List<StoryDetails>>> getStories(String token, int page) async {
    final response = await _dio.get(
      '/story/list',
      queryParameters: {'page': page},
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<StoryDetails>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => StoryDetails.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Belirli bir hikayenin toplam görüntülenme sayısı
  Future<ResponseMessage> getStoryViewCount(String token, String storyId) async {
    final response = await _dio.get(
      '/story/$storyId/views',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Popüler hikayeleri getirme
  Future<DataResponseMessage<List<StoryDTO>>> getPopularStories(String token) async {
    final response = await _dio.get(
      '/story/popular',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<StoryDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => StoryDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Belirli bir hikayeye yapılan yorumları listeleme
  Future<DataResponseMessage<List<CommentDetailsDTO>>> getStoryComments(String token, String storyId) async {
    final response = await _dio.get(
      '/story/$storyId/comments',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<CommentDetailsDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => CommentDetailsDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Belirli bir hikayeye yapılan görüntüleme işlemi
  Future<ResponseMessage> viewStory(String token, String storyId) async {
    final response = await _dio.post(
      '/story/$storyId/view',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Arşivlenmiş hikayeleri alma
  Future<DataResponseMessage<List<StoryDTO>>> getArchivedStories(String token) async {
    final response = await _dio.get(
      '/story/archivedStories',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<StoryDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => StoryDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Arşivden hikaye silme
  Future<ResponseMessage> deleteArchived(String token, String storyId) async {
    final response = await _dio.delete(
      '/story/$storyId/archivedStory',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Belirli bir hikayenin süresini uzatma
  Future<ResponseMessage> extendStoryDuration(String token, String storyId, int hours) async {
    final response = await _dio.put(
      '/story/$storyId/extend',
      queryParameters: {'hours': hours},
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Belirli bir hikayeyi görüntüleyen kullanıcıların listesi
  Future<DataResponseMessage<List<SearchAccountDTO>>> getStoryViewers(String token, String storyId) async {
    final response = await _dio.get(
      '/story/$storyId/viewers',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<SearchAccountDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => SearchAccountDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Öne çıkarılan hikaye detaylarını alma
  Future<DataResponseMessage<FeatureStoryDTO>> getFeatureId(String token, String featureId) async {
    final response = await _dio.get(
      '/story/feature/$featureId',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<FeatureStoryDTO>.fromJson(
      response.data,
      (data) => FeatureStoryDTO.fromJson(data as Map<String, dynamic>),
    );
  }

  // Belirli bir kullanıcının aktif hikayelerini getirme
  Future<DataResponseMessage<List<StoryDTO>>> getUserActiveStories(String token, int userId) async {
    final response = await _dio.get(
      '/story/user/$userId/active',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return DataResponseMessage<List<StoryDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => StoryDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Hikaye beğeni bilgisini alma
  Future<ResponseMessage> getLike(String token, String storyId) async {
    final response = await _dio.get(
      '/story/$storyId/getLike',
      options: Options(headers: {'Authorization': 'Bearer $token'}),
    );
    return ResponseMessage.fromJson(response.data);
  }
}
